### PR TITLE
Getting thrown into things actually hurts now [ready for review]

### PR DESCRIPTION
**Pull request by Qwertytoforty**
## What Does This PR Do

Carbon mobs hitting things now calls a carbon_throw_hit on whatever was hit by a carbon mob. This lets us adjust how things react to being hit by a skrell, and adjust how much damage said mob takes.

Being thrown into another mob / structure actually deals damage now, vs being hit into a wall / another carbon being the only thing that hurts, and hitting like a door or window would deal zero damage.

Objects / mobs that are hit now  take damage from carbons hitting them. Structures generally hurt less than walls

Certain things have been given special effects on hit. Currently that is at the moment:

Windows, hits a person with a shard of glass that can embed if the window shatters from damage.

Vendors, tip in the same way as vendor shoves work currently.

Blob tiles deal less damage / take less damage from thrown carbons, because sorium exists, and should hurt, but should also not cause the blob to rip itself apart / hurt too much

Consoles with their weak glass screens will often break when a person is thrown into them.

## Why It's Good For The Game

Being thrown into something made of metal should hurt. Especially if it is sharp or something. Things should be damaged from being hit by someone power fisted into it too.

## Images of changes

Will get some action clips here

## Testing

throwing carbons at things, see what happens

## TODO

- [x] Improve the proc
- [x] make some other things have special hit effects that should have them
- [x] improve the cl as we expand
- [x] make sure shadow demon doesnt die pulling people at itself

## Changelog
:cl:
tweak: Being thrown into structures / basic mobs hurts now.
tweak: Some structures, like consoles or windows or vendors, will hurt more.
tweak: You can throw people dramatically through damaged windows into space.
/:cl:

### CODE REVIEW

File: code/modules/events/blob/theblob.dm [hunk 1]
```diff
@@ -268,6 +268,10 @@ GLOBAL_LIST_EMPTY(blob_nodes)
 			return B.blob_reagent_datum.description
 	return "something unknown"
 
+/obj/structure/blob/carbon_throw_hit(mob/living/carbon/human/C, datum/thrownthing/throwingdatum, damage, mob_hurt, self_hurt)
+	damage *= 0.25 // Lets not have sorium be too much of a blender / rapidly kill itself
```
Comment (Sirryan2002) [inline code/modules/events/blob/theblob.dm:272]: will this properly be tracked down to the parent procs?

File: code/modules/mob/living/carbon/carbon.dm [hunk 2]
```diff
@@ -612,31 +612,19 @@ GLOBAL_LIST_INIT(ventcrawl_machinery, list(/obj/machinery/atmospherics/unary/ven
...
 		return
 
-	var/hurt = TRUE
 	var/damage = 10 + 1.5 * speed // speed while thrower is standing still is 2, while walking with an aggressive grab is 2.4, highest speed is 14
 	/*if(istype(throwingdatum, /datum/thrownthing))
 		var/datum/thrownthing/D = throwingdatum
 		if(isrobot(D.thrower))
 			var/mob/living/silicon/robot/R = D.thrower
 			if(!R.emagged)
 				hurt = FALSE*/
```
Comment (Sirryan2002) [inline code/modules/mob/living/carbon/carbon.dm:621]:
```suggestion
```
might as well nuke this commented out code

File: code/modules/mob/living/living.dm [hunk 3]
```diff
@@ -1139,3 +1139,14 @@
 /mob/living/throw_at(atom/target, range, speed, mob/thrower, spin, diagonals_first, datum/callback/callback, force, dodgeable)
 	stop_pulling()
 	return ..()
+
+/mob/living/carbon_throw_hit(mob/living/carbon/human/C, datum/thrownthing/throwingdatum, damage, mob_hurt, self_hurt)
+	if(C == src || flying || !density)
+		return
+	if(!mob_hurt)
```
Comment (Sirryan2002) [inline code/modules/mob/living/living.dm:1146]: make this an early return

File: code/game/turfs/turf.dm [hunk 4]
```diff
@@ -573,3 +573,11 @@
 		else
 			return I
 	return I
+
+
+/turf/carbon_throw_hit(mob/living/carbon/human/C, datum/thrownthing/throwingdatum, damage, mob_hurt, self_hurt)
+	if(!mob_hurt && density)
```
Comment (Sirryan2002) [inline code/game/turfs/turf.dm:579]: make this an early return

File: code/game/objects/objs.dm [hunk 5]
```diff
@@ -297,3 +297,12 @@
 	// In the event that the object doesn't have an overriden version of this proc to do it, log a runtime so one can be added.
 	CRASH("Proc force_eject_occupant() is not overridden on a machine containing a mob.")
 
+/obj/carbon_throw_hit(mob/living/carbon/human/C, datum/thrownthing/throwingdatum, damage, mob_hurt, self_hurt)
+	damage *= 0.75 //Define this probably somewhere, we want objects to hurt less than walls, unless special impact effects.
+	if(!mob_hurt && density)
```
Comment (Sirryan2002) [inline code/game/objects/objs.dm:302]: make this an early return

File: code/game/machinery/vendors/vending.dm [hunk 6]
```diff
@@ -1118,6 +1118,14 @@
 	tilt(target, from_combat = TRUE)
 	return TRUE
 
+/obj/machinery/economy/vending/carbon_throw_hit(mob/living/carbon/human/C, datum/thrownthing/throwingdatum, damage, mob_hurt, self_hurt)
+	if(HAS_TRAIT(C, TRAIT_FLATTENED))
+		return
```
Comment (Sirryan2002) [inline code/game/machinery/vendors/vending.dm:1123]: do you just want to return here and null out any effects or just prevent tilt'ing?

File: code/game/machinery/computer/computer.dm [hunk 7]
```diff
@@ -176,3 +176,11 @@
 	if(circuit && !(flags & NODECONSTRUCT))
 		if(I.use_tool(src, user, 20, volume = I.tool_volume))
 			deconstruct(TRUE, user)
+
+/obj/machinery/computer/carbon_throw_hit(mob/living/carbon/human/C, datum/thrownthing/throwingdatum, damage, mob_hurt, self_hurt)
+	if(!self_hurt && prob(50 * damage / 15))
```
Comment (Sirryan2002) [inline code/game/machinery/computer/computer.dm:181]:
```suggestion
	if(!self_hurt && prob(50 * (damage / 15)))
```
properly enclosing fractions from other operations

File: code/game/machinery/vendors/vending.dm (hunk 6 above)
Comment (Qwertytoforty) [reply code/game/machinery/vendors/vending.dm:1123]: Probably should only block tilting

Comment (Sirryan2002) [reply code/game/machinery/vendors/vending.dm:1123]: can just do `return ..()` then

File: code/modules/events/blob/theblob.dm (hunk 1 above)
Comment (Qwertytoforty) [reply code/modules/events/blob/theblob.dm:272]: Should be, but will double check to be 100% sure

File: code/game/machinery/vendors/vending.dm (hunk 6 above)
Comment (Qwertytoforty) [reply code/game/machinery/vendors/vending.dm]: done

File: code/game/objects/objs.dm (hunk 5 above)
Comment (Qwertytoforty) [reply code/game/objects/objs.dm]: done

File: code/game/turfs/turf.dm (hunk 4 above)
Comment (Qwertytoforty) [reply code/game/turfs/turf.dm]: done

File: code/modules/mob/living/living.dm (hunk 3 above)
Comment (Qwertytoforty) [reply code/modules/mob/living/living.dm]: done

File: code/modules/events/blob/theblob.dm (hunk 1 above)
Comment (Qwertytoforty) [reply code/modules/events/blob/theblob.dm:272]: Works (once I made a change as blob tiles are not dense, naturally :gatto: )

**Review by Sirryan2002 — APPROVED** [review]
👏

**Review by S34NW — COMMENTED** [review]
`carbon_throw_hit` sorta gives me the ick, can it be more descriptive? `hit_by_thrown_carbon` or something?

File: code/modules/events/blob/theblob.dm (hunk 1 above)
Comment (S34NW) [inline code/modules/events/blob/theblob.dm:272]: is this intended to stack with the `obj` level multiplication of 0.75?

Comment (Qwertytoforty) [reply code/modules/events/blob/theblob.dm:272]: Yes. More for the knockdown over damage if it goes into another blob, otherwise it will rip itself apart

**Comment by Qwertytoforty** [comment]
> `carbon_throw_hit` sorta gives me the ick, can it be more descriptive? `hit_by_thrown_carbon` or something?

Done!

**Review by S34NW — APPROVED** [review]

**Pull request merged**